cirq/contrib/quirk/cells/arithmetic_cells.py

```diff
@@ -0,0 +1,308 @@
+                                 f'Target: {target}\n'
+                                 f'Modulus: {r}')
+
+        self.identifier = identifier
+        self.target = target
+        self.inputs = inputs
+        self.operation = operation
+
+    def _value_equality_values_(self):
+        return self.identifier, self.operation, self.target, self.inputs
+
+    def registers(self):
+        return [self.target, *self.inputs]
+
+    def with_registers(self, *new_registers: Union[int, Sequence['cirq.Qid']]
+                      ) -> 'QuirkArithmeticOperation':
+        if len(new_registers) != len(self.inputs) + 1:
+            raise ValueError('Wrong number of registers.\n'
+                             f'New registers: {repr(new_registers)}\n'
+                             f'Operation: {repr(self)}')
+
+        if isinstance(new_registers[0], int):
+            raise ValueError('The first register is the mutable target. '
+                             'It must be a list of qubits, not the constant '
+                             f'{new_registers[0]}.')
+
+        return QuirkArithmeticOperation(self.identifier, self.operation,
+                                        new_registers[0], new_registers[1:])
+
+    def apply(self, *registers: int) -> Union[int, Iterable[int]]:
+        return self.operation(*registers)
+
+    def _circuit_diagram_info_(self, args: 'cirq.CircuitDiagramInfoArgs'):
+        lettered_args = list(zip(self.operation.letters, self.inputs))
+
+        result = []
+
+        # Target register labels.
+        consts = ''.join(f',{letter}={reg}' for letter, reg in lettered_args
+                         if isinstance(reg, int))
+        result.append(f'Quirk({self.identifier}{consts})')
+        result.extend(f'#{i}' for i in range(2, len(self.target) + 1))
+
+        # Input register labels.
+        for letter, reg in lettered_args:
+            if not isinstance(reg, int):
+                result.extend(f'{letter.upper()}{i}'
+                              for i in range(len(cast(Sequence, reg))))
+
+        return result
+
+    def __repr__(self):
+        return ('cirq.contrib.quirk.QuirkArithmeticOperation(\n'
+                f'    {repr(self.identifier)},\n'
+                f'    operation={repr(self.operation)},\n'
+                f'    target={repr(self.target)},\n'
+                f'    inputs={_indented_list_lines_repr(self.inputs)},\n'
+                ')')
+
+
+@value.value_equality
+class QuirkArithmeticLambda:
+    """A callable with parameter-name-dependent behavior."""
+
+    def __init__(self, code: str):
+        """
+        Args:
+            code: A string containing a python lambda expression using only
+                values built into python or cirq. The lambda expression is
+                passed as a string to guarantee it has a `repr`.
+        """
+        self.code = code
+        self._func: Callable = eval(code, {'cirq': cirq}, {})
```

Comment:
It is very dangerous to allow input to be interpreted as code. Can we not do that and instead implement the functionality we actually need here? Perhaps we can simply take a lambda directly, rather than taking its string form.

To make this more concrete and entertaining, here's how to delete files:

```python
payload = "__import__('os').system('rm -rf /path/to/victims/files')"
cirq.contrib.quirk.QuirkArithmeticLambda(payload)
```

Reply:
I don't see how the attack surface is larger than taking a normal callable. With a normal callable you can also trivially delete files. But we don't say that python's `map` function is insecure because of this.

Every string that goes through this method is a string literal in cirq. There is no situation where these strings are constructed by pasting together bits of other strings. Well, actually, in tests this does happen in the sense that you call the repr and then eval it. But calling eval inside the eval isn't any *more* problematic.

If an attacker modifies the repr to include a bad string argument here then they can do bad things. But in that case the user already pasted a string into their python interpreter from an attacker. The attacker could just directly delete files instead of going through the roundabout method of involving this file.

Do you know of a situation where this causes a problem that wouldn't also be present for a raw Callable?

Reply:
Use of `eval` here is dangerous and unnecessary.

It's dangerous, because even if you have taken sufficient care to ensure all strings that hide python code are literals today, we have no mechanism to ensure this remains true in future. Also, anything built on top of this by our users is likely to incur a serious security vulnerability.

It's unnecessary, because if we're only concerned with a finite set of lambda expressions, we can easily just list them in a lookup table (and as a minor bonus we'll recover the ability to type-check the code currently hidden in strings).

Reply:
I switched the implementation to use a global hidden table for the operations.